### PR TITLE
[PROTOCOL-689] Claim interest earned from the tDAI pool for NFT stakers

### PR DESCRIPTION
**Pull request by sypeer**
It is:
- Added functionality to allow the claiming of proportional interest earned from the tDAI pool during Alpha

### CODE REVIEW

File: contracts/lending/ttoken/ITToken.sol [hunk 1]
```diff
@@ -2,13 +2,9 @@
...
 // Contracts
-import {
-    ERC20Upgradeable
-} from "@openzeppelin/contracts-upgradeable/token/ERC20/ERC20Upgradeable.sol";
+import { ERC20Upgradeable } from "@openzeppelin/contracts-upgradeable/token/ERC20/ERC20Upgradeable.sol";
 import { ERC20 } from "@openzeppelin/contracts/token/ERC20/ERC20.sol";
-import {
-    RolesFacet
-} from "../../contexts2/access-control/roles/RolesFacet.sol";
+import { RolesFacet } from "../../contexts2/access-control/roles/RolesFacet.sol";
```
Comment (passabilities) [inline contracts/lending/ttoken/ITToken.sol:7]: please make sure you run `yarn install`. The prettier version should have left the formatting as it was.

File: contracts/nft/NFTFacet.sol [hunk 2]
```diff
@@ -56,4 +56,23 @@ contract NFTFacet is RolesMods {
         // Give the caller authorization to protocol
         RolesLib.grantRole(AUTHORIZED, msg.sender);
     }
+
+    /**
+     * @notice It returns the claimable interest % for a user's staked NFTs.
+     * @param nftOwner The address of the NFT owner.
+     */
+    function getClaimableInterestPercent(address nftOwner)
```
Comment (passabilities) [inline contracts/nft/NFTFacet.sol:64]: lets create a new facet so that we don't have to redeploy existing logic on this facet

**Review by passabilities — CHANGES_REQUESTED** [review]

File: contracts/nft/NFTFacet.sol (hunk 2 above)
Comment (sypeer) [reply contracts/nft/NFTFacet.sol]: done

File: contracts/nft/NFTInterestFacet.sol [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+     * @notice It returns the claimable interest % for a user's staked NFTs.
+     * @param nftOwner The address of the NFT owner.
+     */
+    function getClaimableInterestPercent(address nftOwner)
+        external
+        view
+        returns (uint16 claimableInterestPercent)
+    {
+        uint256[] memory stakedNFTs = NFTLib.stakedNFTs(nftOwner);
+        uint256[] memory diamondNFTs = NFTLib.nft().getOwnedTokens(
```
Comment (passabilities) [inline contracts/nft/NFTInterestFacet.sol:22]: there is a slightly easier way to do this:

```
uint256 userNFTs = EnumerableSet.length(
    NFTLib.s().stakedNFTs[nftOwner]
);
uint256 diamondNFTs = NFTLib.nft().balanceOf(address(this));
```

Comment (passabilities) [reply contracts/nft/NFTInterestFacet.sol:22]: that way don't have to copy token IDs into memory array and then only use the `length` value.

File: contracts/nft/NFTInterestFacet.sol [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+pragma solidity ^0.8.0;
+
+// Contracts
+import { NFTFacet } from "./NFTFacet.sol";
+
+// Libraries
+import { NFTLib } from "./libraries/NFTLib.sol";
+import { NumbersLib } from "../shared/libraries/NumbersLib.sol";
+
+contract NFTInterestFacet is NFTFacet {
```
Comment (passabilities) [inline contracts/nft/NFTInterestFacet.sol:11]: we should add a function here that calls the TToken to claim the interest on behalf of the caller. as the claim interest is protected to only be callable from the Diamond 

it should calculate the percentage to use and then call the function you made in `TToken_V3` `claimAlphaInterest` that forwards the callers address and the calculated percentage.

**Review by passabilities — CHANGES_REQUESTED** [review]

File: contracts/nft/NFTInterestFacet.sol (hunk 3 above)
Comment (sypeer) [reply contracts/nft/NFTInterestFacet.sol:22]: This is what I was asking about last week. Will add it in

File: contracts/nft/NFTInterestFacet.sol (hunk 4 above)
Comment (sypeer) [reply contracts/nft/NFTInterestFacet.sol:11]: This is what I was asking about last week. Will add it in

**Review by admazzola — DISMISSED** [review]

**Review by mkhoury93 — APPROVED** [review]

**Review by mkhoury93 — DISMISSED** [review]

File: contracts/lending/ttoken/TToken_V2_Alpha.sol [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    /**
+     * @notice Called by the Teller Diamond contract at the end of Alpha for recipient's earned interest.
+     * @param recipient The address of the alpha participant claiming the interest.
+     * @param percent The proportional amount of interest earned by the recipient.
+     */
+
+    function claimAlphaInterest(address recipient, uint16 percent)
+        external
+        authorized(CONTROLLER, _msgSender())
+    {
```
Comment (passabilities) [inline contracts/lending/ttoken/TToken_V2_Alpha.sol:53]: Should perform a check that ensures `percent > 0`

File: test/integration/claim-alpha-interest.test.ts [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    })
+
+    it('Should not be able to claim interest twice', async () => {
+      const { signer } = await hre.evm.impersonate(staker)
+
+      await diamond
+        .connect(signer)
+        .claimNFTInterest(lendingToken.address)
+        .should.be.revertedWith('Teller: interest already claimed')
+    })
```
Comment (passabilities) [inline test/integration/claim-alpha-interest.test.ts:95]: should add a test that ensures claimable `percent > 0`

**Review by passabilities — CHANGES_REQUESTED** [review]

File: contracts/lending/ttoken/TToken_V2_Alpha.sol (hunk 5 above)
Comment (sypeer) [reply contracts/lending/ttoken/TToken_V2_Alpha.sol:53]: added

File: test/integration/claim-alpha-interest.test.ts (hunk 6 above)
Comment (sypeer) [reply test/integration/claim-alpha-interest.test.ts:104]: L74

**Review by passabilities — DISMISSED** [review]

**Review by mkhoury93 — DISMISSED** [review]

**Review by passabilities — APPROVED** [review]

**Review by dumebi — APPROVED** [review]

**Pull request merged**